src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,131 @@
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterOperationAction(OnCompilationStart, OperationKind.Invocation);
+        }
+
+        private static void OnCompilationStart(OperationAnalysisContext context)
+        {
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? stream) ||
+                stream == null)
+            {
+                return;
+            }
+
+            // Verify that the required Memory types were shipped in the current .NET version, otherwise we should not suggest the overload
+            if (context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemoryExtensions) == null ||
+                context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1) == null ||
+                context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlyMemory1) == null)
+            {
+                return;
+            }
+
+            if (context.Operation is IInvocationOperation invocation &&
+                invocation.TargetMethod is IMethodSymbol method &&
+                IsStreamMethod(method, stream) &&
+                HasUndesiredArguments(method))
+            {
+                if (string.Equals(method.Name, "ReadAsync", StringComparison.Ordinal))
+                {
+                    context.ReportDiagnostic(invocation.Syntax.CreateDiagnostic(PreferStreamReadAsyncMemoryOverloadsRule));
+                }
+                else if (string.Equals(method.Name, "WriteAsync", StringComparison.Ordinal))
+                {
+                    context.ReportDiagnostic(invocation.Syntax.CreateDiagnostic(PreferStreamWriteAsyncMemoryOverloadsRule));
+                }
+            }
+        }
+
+        private static bool IsStreamMethod(IMethodSymbol method, INamedTypeSymbol streamSymbol)
+        {
+            return method.ContainingType.Equals(streamSymbol) ||
+                (method.OverriddenMethod != null && method.OverriddenMethod.ContainingType.Equals(streamSymbol));
+        }
+
+        private static bool HasUndesiredArguments(IMethodSymbol method)
+        {
+            // Both undesired ReadAsync/WriteAsync overloads have the same 3 args and an optional 4th (CancellationToken)
+            return method.Parameters.Length >= 3 &&
+                method.Parameters[0].Type.TypeKind == TypeKind.Array &&
```

Comment:
This line seems redundant given the check in the following line.

Reply:
Removing.